---
title: "AngII_1"
author: "Paul Cho"
date: "2025-12-19"
output:
  html_document:
    toc: true
    toc_depth: 4
    toc_float:
      collapsed: true
      number_sections: true
    theme: journal
    df_print: paged
    code_folding: hide
    highlight: pygments
  word_document:
    toc: true
    toc_depth: '4'
editor_options:
  chunk_output_type: inline
---

# Loading packages
```{r}
options(repos = c(CRAN = "https://cloud.r-project.org"))
if (!requireNamespace("remotes", quietly = TRUE)) {
  install.packages("remotes")
}
library(remotes)
```

```{r}
if (!require("here")) {install.packages("here"); require("here")}

Datasets <- "Datasets"
Outputs <- "Outputs"

if (!dir.exists(here(Datasets))) {dir.create(here(Datasets))}
if (!dir.exists(here(Outputs))) {dir.create(here(Outputs))}
```


```{r}
install.packages("remotes")
remotes::install_github('chris-mcginnis-ucsf/DoubletFinder', force = TRUE)

install.packages("devtools")
devtools::install_github("davidsjoberg/ggsankey")
```

```{r}
if (!require("dplyr")) {install.packages("dplyr"); require("dplyr")}
if (!require("Seurat")) {install.packages("Seurat"); require("Seurat")}
if (!require("patchwork")) {install.packages("patchwork"); require("patchwork")}
if (!require("cowplot")) {install.packages("cowplot"); require("cowplot")}
if (!require("ggpubr")) {install.packages("ggpubr"); require("ggpubr")}
if (!require("plotly")) {install.packages("plotly"); require("plotly")}
if (!require("knitr")) {install.packages("knitr"); require("knitr")}
if (!require("htmlwidgets")) {install.packages("htmlwidgets"); require("htmlwidgets")}
if (!require("tidyverse")) {install.packages("tidyverse"); require("tidyverse")} # for titying up data
if (!require("RColorBrewer")) {install.packages("RColorBrewer"); require("RColorBrewer")} # for color brewer
if (!require("sctransform")) {install.packages("sctransform"); require("sctransform")} # for data normalization
if (!require("openxlsx")) {install.packages("openxlsx"); require("openxlsx")} # to save .xlsx files
if (!require("SoupX")) {install.packages("SoupX"); require("SoupX")}
if (!require("readxl")) {install.packages("readxl"); require("readxl")}
if (!require("hdf5r")) {install.packages("hdf5r"); require("hdf5r")}

if (!require("glmGamPoi")) {BiocManager::install('glmGamPoi'); require("glmGamPoi")} # for data normalization, sctransform
if (!require("EnhancedVolcano")) {BiocManager::install('EnhancedVolcano'); require("EnhancedVolcano")} # volcano plot

library(ggsankey)
library(DoubletFinder)

set.seed((12345))
here()

#renv::init()
#renv::install("Seurat@5.2.0")
#renv::install("hdf5r")
#renv::install("chris-mcginnis-ucsf/DoubletFinder")
#renv::snapshot()
```

# Loading the dataset

```{r}
kidney1 <- "kidney1"
```

```{r}
# Create a Soup Channel Object(sc) to use for future filtering of ambient RNA
RawDataset = Read10X_h5(here("datasets", "1_raw_feature_bc_matrix.h5"))
FilteredDataset = Read10X_h5(here("datasets", "1_filtered_feature_bc_matrix.h5"))

# Make a Seurat Object from the filtered control data 
filtered_kidney <- Read10X_h5(here("datasets", "1_filtered_feature_bc_matrix.h5"))
filtered_kidney <- CreateSeuratObject(counts = filtered_kidney, project = "Filteredkidney_snRNA")
```

#SoupX

```{r}
sc = SoupChannel(RawDataset,FilteredDataset)

# Cluster the Cells with Seurat 
filtered_kidney <- SCTransform(filtered_kidney, verbose = F)
filtered_kidney <- RunPCA(filtered_kidney, verbose = F)
filtered_kidney <- RunUMAP(filtered_kidney, dims = 1:30, verbose = F)
filtered_kidney <- FindNeighbors(filtered_kidney, dims = 1:30, verbose = F)
filtered_kidney <- FindClusters(filtered_kidney, verbose = F)

meta <- filtered_kidney@meta.data
umap <- filtered_kidney@reductions$umap@cell.embeddings
clusters <- setNames(meta$seurat_clusters, rownames(meta))
```


```{r}
#Ensure the two following numbers are equal before continuing 
check1 <- length(clusters)
check2 <- nrow(sc$metaData)

sc <- setClusters(sc, clusters)
sc <- setDR(sc, umap)
```

Calculating contamination fraction:
```{r}
sc = autoEstCont(sc)
```
Post rho value is 1.7% (estimated amount of ambient RNA is 1.7%, low)

```{r}
#Clean the data
filtered_kidney_out = adjustCounts(sc)
 
#Create a new Seurat Object out of the cleaned data
seurat.obj <- CreateSeuratObject(filtered_kidney_out)

# Remove unused objects in environment to save RAM

#rm(filtered_kidney, filtered_kidney_out, FilteredDataset, RawDataset, sc, meta, check1, check2, clusters, umap, y, Datasets, Outputs)
```

# Doublet filtering

1. DoubletFinder: removing doublets(droplets that captured 2+ cells instead of 1)
- After running, look at "DF.class", which labels each cell as “Doublet” or “Singlet.”
- Visualize results using UMAP.
- pANN score: how doublet-like a cell is. Useful for threshold tuning.
2. Manual doublet curation

```{r}
Start <- ncol(seurat.obj)

p1 <- VlnPlot(seurat.obj, features = c("nFeature_RNA")) + theme(axis.title.x = element_blank()) + theme(legend.position = 'none') + theme(axis.text.x = element_text(angle = 0, hjust = .5))

p2 <- VlnPlot(seurat.obj, features = c("nFeature_RNA"), pt.size = 0) + theme(axis.title.x = element_blank()) + theme(legend.position = 'none') + theme(axis.text.x = element_text(angle = 0, hjust = .5))

p3 <- VlnPlot(seurat.obj, features = c("nCount_RNA")) + theme(axis.title.x = element_blank())  + theme(legend.position = 'none') + theme(axis.text.x = element_text(angle = 0, hjust = .5))

p4 <- VlnPlot(seurat.obj, features = c("nCount_RNA"), pt.size = 0) + theme(axis.title.x = element_blank())  + theme(legend.position = 'none') + theme(axis.text.x = element_text(angle = 0, hjust = .5))

cowplot::plot_grid(p1, p2, p3, p4, ncol = 4, scale = .85) + draw_label("Before Filtering", y = .95, size = 22, fontface = "bold")

```
```{r}
# Minimal filtering of low quality cells 
seurat.obj.f <- subset(seurat.obj, nFeature_RNA > 1000)

p1 <- VlnPlot(seurat.obj.f, features = c("nFeature_RNA")) + theme(axis.title.x = element_blank()) + theme(legend.position = 'none') + theme(axis.text.x = element_text(angle = 0, hjust = .5))

p2 <- VlnPlot(seurat.obj.f, features = c("nFeature_RNA"), pt.size = 0) + theme(axis.title.x = element_blank()) + theme(legend.position = 'none') + theme(axis.text.x = element_text(angle = 0, hjust = .5))

p3 <- VlnPlot(seurat.obj.f, features = c("nCount_RNA")) + theme(axis.title.x = element_blank())  + theme(legend.position = 'none') + theme(axis.text.x = element_text(angle = 0, hjust = .5))

p4 <- VlnPlot(seurat.obj.f, features = c("nCount_RNA"), pt.size = 0) + theme(axis.title.x = element_blank())  + theme(legend.position = 'none') + theme(axis.text.x = element_text(angle = 0, hjust = .5))

cowplot::plot_grid(p1, p2, p3, p4, ncol = 4, scale = .85) + draw_label("After Filtering Low Complexity Cells", y = .95, size = 22, fontface = "bold")
```

```{r}
Filter_1 <- ncol(seurat.obj.f)

amountfiltered1 <- Start-Filter_1

#Making a table to display the amount of cells being filtered out with the minimal filtering of low quality cells applied to the snRNAseq dataset. 
cell_counts <- data.frame(Phase = c("Before Filtering", "After Filtering", "Amount Filtered"), Cell_Number = c(Start, Filter_1, amountfiltered1))

print(cell_counts)

# Remove unused objects in environment to save RAM

#rm(seurat.obj, p1, p2, p3, p4, cell_counts)
```

# High-dimensional superclustering

```{r}
resolution_value <- 1.5
```

```{r}
# Pre-process standard workflow
seurat.obj.f <- NormalizeData(object = seurat.obj.f)
seurat.obj.f <- FindVariableFeatures(object = seurat.obj.f)
seurat.obj.f <- ScaleData(object = seurat.obj.f)
seurat.obj.f <- RunPCA(object = seurat.obj.f)

# ElbowPlot(seurat.obj.f, ndims = 30)

seurat.obj.f <- FindNeighbors(object = seurat.obj.f, dims = 1:30)
seurat.obj.f <- FindClusters(object = seurat.obj.f, resolution = resolution_value, verbose = F)

# Make a UMAP to visualize the clusters 
seurat.obj.f <- RunUMAP(object = seurat.obj.f, dims = 1:30)
DimPlot(seurat.obj.f, reduction = "umap", label = TRUE) +
   theme(plot.title = element_text(hjust = 0.5)) +
   ggtitle(paste0("Initial Clustering for ", kidney1))
```

# Save excel spreadsheet of DEG List by Cluster
This file can be used to determine the prominent features (highly enriched genes) of each cluster. + will be in the Outputs folder

```{r}
all_markers <- FindAllMarkers(seurat.obj.f, only.pos = TRUE, min.pct = 0.5, logfc.threshold = 1)

# Order by avg_log2FC
all_markers <- all_markers %>%
  arrange(desc(avg_log2FC)) %>%
  select(gene, everything())

# Split by cluster (ident column)
marker_list <- split(all_markers, all_markers$cluster)

# Sort names alphanumerically
sorted_cluster_names <- names(marker_list)[order(as.numeric(names(marker_list)))]

# Create a workbook
wb <- createWorkbook()

# Add each cluster as a new worksheet
for (cluster_name in sorted_cluster_names) {
  addWorksheet(wb, sheetName = cluster_name)
  writeData(wb, sheet = cluster_name, marker_list[[cluster_name]])
}

date <- format(Sys.Date(), "%Y%m%d")

# Save workbook
saveWorkbook(wb, here("Outputs", paste0(date, "_", kidney1, "_FindAllMarkers_By_Cluster.xlsx")), overwrite = TRUE)
```

```{r}
cell_type_markers_df <- data.frame(
  Subclass = c(
    "PTS1", "PTS2", "PTS3", "DTL", "TALA", "TALB", "MD", "DCT1", "DCT2", "CNT",
    "PC", "ICA", "ICB", "PODO", "PEC", "URO", "EC", "LYMPH", "FIB", "MES", "VSMC", "LYMPHO", "MACRO"
  ),
  Positive_Marker_1 = c(
    "Lrp2", "Lrp2", "Lrp2", "Epha7", "Slc12a1", "Slc12a1", "Nos1", "Slc12a3", "Slc12a3", "Slc8a1",
    "Aqp2", "Slc4a1", "Slc26a4", "Nphs1", "Ncam1", "Upk1b", "Flt1", "Lyve1", "Pdgfra", "Pdgfrb", "Pdgfrb", "Ptprc", "Ptprc"
  ),
  Positive_Marker_2 = c(
    "Slc5a12", "Slc13a3", "Slc16a9", "Cryab", "Cldn16", "Cldn10", "Slc12a1", "Egf", "Slc8a1", NA,
    NA, "Kit", NA, NA, NA, NA, "Emcn", "Kdr", "Pdgfrb", "Piezo2", "Acta2", "Skap1", "Cd74"
  ),
  stringsAsFactors = FALSE
)

# Creating a dot plot to cross reference the features with the cell clusters present, seeing the average expression and percent expressed 

markers.to.plot1 <- c("Lrp2",         # PTS1, PTS2, PTS3
                      "Slc5a12",      # PTS1
                      "Slc13a3",      # PTS2
                      "Slc16a9",      # PTS3
                      "Epha7",        # DTL
                      "Cryab",        # DTL
                      "Slc12a1",      # TALA, TALB, MD
                      "Cldn16",       # TALA
                      "Cldn10",       # TALB
                      "Nos1",         # MD
                      "Slc12a3",      # DCT1, DCT2
                      "Egf",          # DCT1
                      "Slc8a1",       # CNT, DCT2
                      "Aqp2",         # PC
                      "Slc4a1",       # ICA
                      "Kit",          # ICA
                      "Slc26a4",      # ICB
                      "Nphs1",        # PODO
                      "Ncam1",        # PEC
                      "Upk1b",        # URO
                      "Flt1",         # EC
                      "Emcn",         # EC
                      "Lyve1",        # LYMPH
                      "Kdr",          # LYMPH
                      "Pdgfra",       # FIB
                      "Pdgfrb",       # FIB, MES, VSMC
                      "Piezo2",       # MES
                      "Acta2",        # VSMC
                      "Ptprc",        # LYMPHO, MACRO
                      "Skap1",        # LYMPHO
                      "Cd74"          # MACRO
)

p1 <- DotPlot(seurat.obj.f,
features = markers.to.plot1,
dot.scale = 5,
dot.min = 0,
scale.max = 100,
scale.min = 0,
col.min = -2.5,
col.max = 2.5) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  coord_flip() +
  labs(x = "Gene Features", y = "Clusters") +
  theme(
    axis.text.x  = element_text(size = 8, angle = 60, hjust = 1, vjust = 1),
    axis.text.y  = element_text(size = 8),                                   
  )

p1
```

# Missing cell types
```{r}
p2 <- as.data.frame(p1$data)

low_pct_features <- p2 %>%
  group_by(features.plot) %>%
  filter(all(pct.exp <= 75)) %>%
  distinct(features.plot) %>%
  pull(features.plot)

# low_pct_features 

low_pct_features2 <- as.data.frame(low_pct_features)

markers_long <- cell_type_markers_df %>%
  pivot_longer(cols = c(Positive_Marker_1, Positive_Marker_2),
               names_to = "marker_type", values_to = "marker") %>%
  filter(!is.na(marker))  # remove any NAs in marker

# Step 2: Join with low_pct_features2
low_pct_features2_annotated <- low_pct_features2 %>%
  left_join(markers_long, by = c("low_pct_features" = "marker")) %>%
  rename(cell_type = Subclass)  # rename Subclass to cell_type

missing_cell_types <- unique(na.omit(low_pct_features2_annotated$cell_type))

missing_cell_types
```

The following features were found to have less than 75% of the cells expressing them in the any clusters ("DTL"    "TALA"   "MD"     "ICA"    "URO"    "LYMPH"  "MES"    "VSMC"   "LYMPHO")


See if there are any clusters that are clearly enriched for the cell-type defining markers:
```{r}
for (i in low_pct_features) {
  
  # Find the subclass (if any) where the marker matches Positive_Marker_1 or _2
  subclass_match <- cell_type_markers_df %>%
    filter(Positive_Marker_1 == i | Positive_Marker_2 == i) %>%
    pull(Subclass)

  # If found, use the first match; else NA
  subclass_label <- if (length(subclass_match) > 0) subclass_match[1] else NA
  
  # Build the title
  title_text <- if (!is.na(subclass_label)) {
    paste0(i, " (", subclass_label, ")")
  } else {
    i
  }
  
  # Create plot
  p <- FeaturePlot(seurat.obj.f, features = i) &
       scale_colour_gradientn(colours = rev(brewer.pal(n = 9, name = "RdYlBu"))) &
       ggtitle(title_text)
  
  print(p)  # Or save/store as needed
}
```
- Signals more scattered = ambient RNA.
- Should be near where biology expects.

Upk1b(URO), Piezo2 (MES)

# Continue with DoubletFinder
```{r}
# PCs set to 1:20 since it was between 15-20 on the earlier elbow plot 
sweep.res.list_seurat.obj.f <- paramSweep(seurat.obj.f, PCs = 1:20, sct = FALSE) 
```

```{r}
#Summarize each combination of pN and pK
sweep.stats_seurat.obj.f <- summarizeSweep(sweep.res.list_seurat.obj.f, GT = FALSE) 

#Select the pK that corresponds to max bcmvn to optimize doublet detection
bcmvn_seurat.obj.f <- find.pK(sweep.stats_seurat.obj.f)
```

```{r}
pK <- bcmvn_seurat.obj.f %>% 
  filter(BCmetric == max(BCmetric)) %>%
  select(pK) 

#See pK in the Values Environment
pK <- as.numeric(as.character(pK[[1]]))
```

```{r}
# Homotypic Doublet Proportion Estimate -------------------------------------------------------------------------------------
annotations <- seurat.obj.f@meta.data$seurat_clusters  
 
homotypic.prop <- modelHomotypic(annotations)           

# 10X Multiplet Rate Table (the doublet ratio is # of cells recovered divided by 125000) https://kb.10xgenomics.com/hc/en-us/articles/360001378811-What-is-the-maximum-number-of-cells-that-can-be-profiled-

nExp_poi <- round(nrow(seurat.obj.f@meta.data) # To calculate cell number
                  /125000              # To calculate the doublet ratio
                  *nrow(seurat.obj.f@meta.data))

nExp_poi_adj <- round(nExp_poi*(1-homotypic.prop))
```

```{r}
seurat.obj.f_doublets <- doubletFinder(seurat.obj.f,
                        PCs = 1:20,
                        pN = 0.25,
                        pK = pK,
                        nExp = nExp_poi_adj,
                        reuse.pANN = NULL, sct = FALSE)

# Change names of columns for future analysis 
colnames(seurat.obj.f_doublets@meta.data)[grep("^pANN", colnames(seurat.obj.f_doublets@meta.data))] <- "pANN"
colnames(seurat.obj.f_doublets@meta.data)[grep("^DF.class", colnames(seurat.obj.f_doublets@meta.data))] <- "DF.class"

seurat.obj.f_doublets@meta.data$DF.class <- factor(
  seurat.obj.f_doublets@meta.data$DF.class,
  levels = c("Singlet", "Doublet")
)
table(seurat.obj.f_doublets@meta.data$DF.class)

# Make a value for the doublet count to include in the analysis
doublet_count <- table(seurat.obj.f_doublets@meta.data$DF.class)["Doublet"]
```

```{r}
p1 <- VlnPlot(seurat.obj.f_doublets, features = c("nFeature_RNA"), group.by = "DF.class") +
  theme(axis.title.x = element_blank(),
        legend.position = 'none',
        axis.text.x = element_text(angle = 0, hjust = .5))

p2 <- VlnPlot(seurat.obj.f_doublets, features = c("nCount_RNA"), group.by = "DF.class") +
  theme(axis.title.x = element_blank(),
        legend.position = 'none',
        axis.text.x = element_text(angle = 0, hjust = .5))

p3 <- DimPlot(seurat.obj.f_doublets, group.by = "DF.class")

grid <- plot_grid(p1, p2, p3, ncol = 3, align = "h", axis = "tb", rel_widths = c(1, 1, 2))

plot_grid(
  ggdraw() + draw_label("Violin and Feature Plot by DF.class", fontface = 'bold', size = 22, hjust = 0.5),
  grid,
  ncol = 1,
  rel_heights = c(0.1, 1)
)
```

```{r}
seurat.obj.f_singlets <- subset(seurat.obj.f_doublets, DF.class == "Singlet")

DimPlot(seurat.obj.f_singlets, reduction = "umap", label = T)+
   theme(plot.title = element_text(hjust = 0.5)) +
   ggtitle(paste0("UMAP for ", kidney1, " after DoubletFinder"))
```

```{r}
Filter_2  <- ncol(seurat.obj.f_singlets)

amountfiltered2 <- Filter_1-Filter_2

#Making a table to display the amount of cells being filtered out with the minimal filtering of low quality cells applied to the snRNAseq dataset. 
cell_counts <- data.frame(Phase = c("Before Filtering", "After Filtering", "Amount Filtered"), Cell_Number = c(Filter_1, Filter_2, amountfiltered2))

print(cell_counts)

# Remove unused objects in environment to save RAM
# rm(seurat.obj.f, seurat.obj.f_doublets, sweep.res.list_seurat.obj.f, sweep.stats_seurat.obj.f, p, p1, p2)
```

773 cells were removed after DoubletFinder.

# Manual doublet filtering

```{r}
f1 <- DotPlot(seurat.obj.f_singlets,
features = markers.to.plot1,
dot.scale = 8,
dot.min = 0,
scale.max = 100,
scale.min = 0,
col.min = -2.5,
col.max = 2.5) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  coord_flip() +
  labs(x = "Gene Features", y = "Clusters")

f1
```

# Clusters of concern

```{r}
f2 <- as.data.frame(f1$data)

maybe_doublets <- f2 %>%
  group_by(id) %>%
  filter(all(avg.exp.scaled <= 2.0)) %>%
  distinct(id) %>%
  pull(id)

maybe_doublets

# Create color palette: grey for all, color for maybe_doublets
all_clusters <- levels(seurat.obj.f_singlets)
cluster_colors <- setNames(rep("grey80", length(all_clusters)), all_clusters)
cluster_colors[maybe_doublets] <- "red"

# Get UMAP coordinates and cluster assignments
umap_data <- Embeddings(seurat.obj.f_singlets, "umap") %>%
  as.data.frame() %>%
  mutate(cluster = Idents(seurat.obj.f_singlets))

# Compute cluster centroids
cluster_centroids <- umap_data %>%
  group_by(cluster) %>%
  summarise(UMAP_1 = mean(umap_1), UMAP_2 = mean(umap_2))

# Keep only centroids for maybe_doublets
label_data <- cluster_centroids %>%
  filter(cluster %in% maybe_doublets)

# Plot
DimPlot(seurat.obj.f_singlets, reduction = "umap", cols = cluster_colors) +
  ggrepel::geom_text_repel(
    data = label_data,
    aes(x = UMAP_1, y = UMAP_2, label = cluster),
    size = 4,
    color = "black",
    fontface = "bold",
    max.overlaps = Inf,
    box.padding = 0.5,
    point.padding = 0.5
  )
```

```{r}
# Add doublet cluster numbers like the code below to classify clusters as doublets

doublet_clusters <- c(2, 5, 14, 17, 30)
```

```{r}
# Annotate metadata
seurat.obj.f_singlets@meta.data <- seurat.obj.f_singlets@meta.data %>%
  mutate(subclass.doublet = if_else(
    seurat_clusters %in% doublet_clusters, "DOUBLET", "SINGLET"
  ))

# Change the identities to align with new subclass names and make umap
Idents(seurat.obj.f_singlets) <- seurat.obj.f_singlets@meta.data$subclass.doublet

f1 <- DimPlot(seurat.obj.f_singlets, reduction = "umap") +
  theme(plot.title = element_text(hjust = 0.5)) +
  ggtitle("Before Manual Doublet Filtering")

# Double check the table to ensure SINGLET & DOUBLET cluster appear with this function
table(Idents(seurat.obj.f_singlets))
```

```{r}
# Clean out the doublets manually that were identified 
seurat.obj.f_cleaned <- subset(seurat.obj.f_singlets, idents = "DOUBLET", invert = TRUE)

# Double check that the identity was removed 
table(Idents(seurat.obj.f_cleaned))

# Make umap to double check removal 
f2 <- DimPlot(seurat.obj.f_cleaned, reduction = "umap") +
  theme(plot.title = element_text(hjust = 0.5)) +
  ggtitle("After Manual Doublet Filtering")

f1+f2
```
```{r}
# Remove unused objects in environment to save RAM
rm(seurat.obj.f_singlets)
```

# Cell annotation: Re-cluster singlets

```{r}
seurat.obj.f_cleaned <- SCTransform(seurat.obj.f_cleaned) %>%
    RunPCA() %>%
    FindNeighbors(dims = 1:30) %>%
    FindClusters(resolution = 2, verbose = FALSE) %>%
    RunUMAP(dims = 1:30)

DimPlot(seurat.obj.f_cleaned, reduction = "umap", label = TRUE) +  
  theme(plot.title = element_text(hjust = 0.5)) +
  ggtitle(paste0("Non-annotated Clusters for ", kidney1))
```

```{r}
DefaultAssay(seurat.obj.f_cleaned) <- "RNA"
```

```{r}
# Make mdd for all removed doublets 

DotPlot(seurat.obj.f_cleaned,
features = markers.to.plot1,
dot.scale = 8,
dot.min = 0,
scale.max = 100,
scale.min = 0,
col.min = -2.5,
col.max = 2.5) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  coord_flip() +
  labs(x = "Gene Features", y = "Clusters") +
  theme(
    axis.text.x  = element_text(size = 8, angle = 60, hjust = 1, vjust = 1),
    axis.text.y  = element_text(size = 8),                                   
  )
p1
```

```{r}
# Cluster 0: Slc13a3 => Lrp2 => Slc16a9: PTS2?
# Cluster 1: Slc5a12 > Lrp2: PTS1
# Cluster 2: Slc12a1 > Egf > Cldn10: TALB?
# Cluster 3: Slc16a9 > Lrp2 > Cldn10: PTS3
# Cluster 4: Slc13a3 > Lrp2 > Cldn10 > Slc16a9: PTS2? more likely (0 and 4 likely both PTS2)
# Cluster 5: Flt1 = Emcn > Kdr > Piezo2: EC
# Cluster 6: Slc5a12 > Lrp2: PTS1? 1 and 6 likely both PTS1
# Cluster 7: Slc12a3 = Egf > Slc8a1: DCT1
# Cluster 8: Slc13a3 > Lrp2 > Cldn10 > Slc5a12: PTS2 (0, 4 and 8 all PTS2?)
# Cluster 9: Slc8a1 > Aqp2 > Slc12a3: CNT? or DCT2 due to presence of Slc12a3
# Cluster 10: Slc12a1 = Egf > Cldn10 > Cldn16: TALB? (2 and 10 likely both TALB, or 10 might be TALA)
# Cluster 11: Pdgfra > Pdgfrb: FIB
# Cluster 12: Slc5a12 > Lrp2 > Cldn10: PTS1 (1, 6 & 12 all PTS1?)
# Cluster 13: Aqp2 > Slc8a1: Likely PC
# Cluster 14: Emcn > Fit1 >> Kdr: EC
# Cluster 15: Slc16a9 > Lrp2 > Cldn10: PTS3
# Cluster 16: Epha7 > Cryab >> Ncam1: DTL
# Cluster 17: Slc12a3 > Slc8a1 > Egf: DCT2
# Cluster 18: Cryab > Upk1b > Slc8a1 > Aqp2: DTL? or URO
# Cluster 19: Kit > Slc4a1 > Slc8a1 > Cryab: ICA
# Cluster 20: Cryab > Epha7 > Cldn10: DTL
# Cluster 21: Slc26a4 > Kit: ICB
# Cluster 22: Aqp2 > Cryab > Slc8a1: PC? or DTL? (13 & 22 llikely both PC)
# Cluster 23: Cryab > Aqp2 > Slc8a1: PC? (13, 22 & 23 all PC?)
# Cluster 24: Slc12a1 >> Egf > Cldn10: TALB(2, 10 & 24 all TALB?)
# Cluster 25: Nphs1 > Egf > Ncam1: PODO
# Cluster 26: Slc8a1 > Pdgfrb > Piezo2 > Acta2: CNT? or VSMC
```

```{r}
# Apply names to the clusters by subclass
seurat.obj.f_cleaned@meta.data <- seurat.obj.f_cleaned@meta.data %>% mutate(subclass = dplyr::case_when(
seurat_clusters == 0 ~ "PTS2",
seurat_clusters == 1 ~ "PTS1",
seurat_clusters == 2 ~ "TALB",
seurat_clusters == 3 ~ "PTS3",
seurat_clusters == 4 ~ "PTS2",
seurat_clusters == 5 ~ "EC",
seurat_clusters == 6 ~ "PTS1",
seurat_clusters == 7 ~ "DCT1",
seurat_clusters == 8 ~ "PTS2",
seurat_clusters == 9 ~ "CNT",
seurat_clusters == 10 ~ "TALB",
seurat_clusters == 11 ~ "FIB",
seurat_clusters == 12 ~ "PTS1",
seurat_clusters == 13 ~ "PC",
seurat_clusters == 14 ~ "EC",
seurat_clusters == 15 ~ "PTS3",
seurat_clusters == 16 ~ "DTL",
seurat_clusters == 17 ~ "DCT2",
seurat_clusters == 18 ~ "DTL",
seurat_clusters == 19 ~ "ICA",
seurat_clusters == 20 ~ "DTL",
seurat_clusters == 21 ~ "ICB",
seurat_clusters == 22 ~ "PC",
seurat_clusters == 23 ~ "PC",
seurat_clusters == 24 ~ "TALB",
seurat_clusters == 25 ~ "PODO",
seurat_clusters == 26 ~ "CNT"
))

# Filter out the doublets

seurat.obj.f_cleaned <- subset(seurat.obj.f_cleaned, subset = subclass == "DOUBLET", invert = TRUE)

subclass_order <- c("PTS1", "PTS2", "PTS3", "DTL", "TALA", "TALB", "MD", "DCT1", "DCT2", "CNT", "PC", "ICA", "ICB", "PODO", "PEC", "URO", "EC", "FIB", "MES", "LYMPHO", "MACRO")

seurat.obj.f_cleaned@meta.data$subclass <- factor(seurat.obj.f_cleaned@meta.data$subclass, levels = subclass_order)

# Change the identities to align with new subclass names and make umap
Idents(seurat.obj.f_cleaned) <- seurat.obj.f_cleaned@meta.data$subclass

DimPlot(seurat.obj.f_cleaned, reduction = "umap", label = TRUE, label.size = 4) +  
  theme(plot.title = element_text(hjust = 0.5)) +
  ggtitle(paste0("Subclass Annotation for ", kidney1))
```

### Notes: PTS1/PTS2 together & nearby(?) but distinct PTS3 expresses more distal PT clusters: clear separation shows sufficient sequencing depth
### DCT1 => DCT2 => CNT => PC continuu present
### Where are TALA & MD?
### Shouldn't DCT1 be ~3x the size of DCT2?
### Otherwise looks fine

```{r}
Filter_3  <- ncol(seurat.obj.f_cleaned)

amountfiltered3 <- Filter_2-Filter_3

#Making a table to display the amount of cells being filtered out with the minimal filtering of low quality cells applied to the snRNAseq dataset. 
cell_counts <- data.frame(Phase = c("Before Filtering", "After Filtering", "Amount Filtered"), Cell_Number = c(Filter_2, Filter_3, amountfiltered3))

print(cell_counts)
```

```{r}
# Step 1: Find markers (grouped by subclass)
all_markers <- FindAllMarkers(
  seurat.obj.f_cleaned,
  only.pos = TRUE,
  min.pct = 0.5,
  logfc.threshold = 1,
  group.by = "subclass"  # grouping by subclass instead of cluster ID
)

# Step 2: Sort markers by descending log2FC, put gene column first
all_markers <- all_markers %>%
  arrange(desc(avg_log2FC)) %>%
  select(gene, everything())

# Step 3: Split markers by subclass (stored in 'cluster' column still)
marker_list <- split(all_markers, all_markers$cluster)

# Step 4: Reorder the list according to your subclass_order
# First ensure you're only including subclasses that are actually present
valid_subclasses <- subclass_order[subclass_order %in% names(marker_list)]

# Reorder marker_list based on that
marker_list <- marker_list[valid_subclasses]

# Step 5: Write to Excel
wb <- createWorkbook()

for (subclass in names(marker_list)) {
  addWorksheet(wb, sheetName = subclass)
  writeData(wb, sheet = subclass, marker_list[[subclass]])
}

# Step 6: Save file
date <- format(Sys.Date(), "%Y%m%d")
saveWorkbook(
  wb,
  here("Outputs", paste0(date, "_", kidney1, "_FindAllMarkers_By_Subclass.xlsx")),
  overwrite = TRUE
)
```

```{r}
DimPlot(seurat.obj.f_cleaned, reduction = "umap", label = TRUE, label.size = 4) +  
  theme(plot.title = element_text(hjust = 0.5)) +
  ggtitle(paste0("Subclass Annotation for ", kidney1))
```
# Linear(ish) Dotplot

```{r}
# Change the order of the identities on the x-axis to create a diagonal on the dotplot, going from top to bottom of the "naming kidney cells" table

DotPlot(seurat.obj.f_cleaned,
features = markers.to.plot1,
group.by = "subclass",
dot.scale = 8,
dot.min = 0,
scale.max = 100,
scale.min = 0,
col.min = -2.5,
col.max = 2.5) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  coord_flip() +
  labs(x = "Gene Features", y = "Clusters") +  
  theme(plot.title = element_text(hjust = 0.5)) +
  ggtitle(paste0("Subclass Annotation for ", kidney1))
```

# Annotating by class
```{r}
# Change the specificty of the cells to a lower level and make another UMAP
seurat.obj.f_cleaned@meta.data <- seurat.obj.f_cleaned@meta.data %>%
  mutate(class = dplyr::case_when(
    subclass %in% c("PTS1", "PTS2", "PTS3") ~ "PT",
    subclass == "DTL" ~ "DTL",
    subclass %in% c("TALA", "TALB", "MD") ~ "TAL",
    subclass %in% c("DCT1", "DCT2") ~ "DCT",
    subclass == "CNT" ~ "CNT",
    subclass == "PC" ~ "PC",
    subclass == "ICA" ~ "ICA",
    subclass == "ICB" ~ "ICB",
    subclass == "PODO" ~ "PODO",
    subclass == "PEC" ~ "PEC",
    subclass == "URO" ~ "URO",
    subclass %in% c("EC", "ECG", "ECP", "LYMPH") ~ "EC",
    subclass == "FIB" ~ "FIB",
    subclass == "MES" ~ "MES",
    subclass == "VSMC" ~ "VSMC",
    subclass %in% c("LYMPHO", "MACRO") ~ "IMMUNE",
  ))

DimPlot(seurat.obj.f_cleaned, group.by = "class", label = TRUE, label.size = 4) +
   theme(plot.title = element_text(hjust = 0.5)) +
   ggtitle(paste0("Class Annotation for ", kidney1))
```

```{r}
class_order <- c("PT", "DTL", "TAL", "DCT", "CNT", "PC", "ICA", "ICB", "PODO", "PEC", "URO", "EC", "FIB", "MES", "IMMUNE")
seurat.obj.f_cleaned@meta.data$class <- factor(seurat.obj.f_cleaned@meta.data$class, levels = class_order)

markers.to.plot2 <- c("Lrp2",         # PTS1, PTS2, PTS3
                      "Cryab",        # DTL
                      "Slc12a1",      # TAL1, TAL2, MD
                      "Slc12a3",      # DCT1, DCT2
                      "Slc8a1",       # CNT
                      "Aqp2",         # PC
                      "Kit",          # ICA
                      "Slc26a4",      # ICB
                      "Nphs1",        # PODO
                      "Ncam1",        # PEC
                      "Upk1b",        # URO
                      "Flt1",         # EC
                      "Pdgfra",       # FIB
                      "Pdgfrb",       # FIB, MES, VSMC
                      "Piezo2",       # MES
                      "Ptprc"        # LYMPHO, MACRO
)

DotPlot(seurat.obj.f_cleaned,
features = markers.to.plot2,
group.by = "class",
dot.scale = 8,
dot.min = 0,
scale.max = 100,
scale.min = 0,
col.min = -2.5,
col.max = 2.5) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  coord_flip() +
  labs(x = "Gene Features", y = "Clusters") +  
  theme(plot.title = element_text(hjust = 0.5)) +
  ggtitle(paste0("Class Annotation for ", kidney1))
```

# Annotating by Type

```{r}
# Change the specificty of the cells to a further lower level and make another UMAP
seurat.obj.f_cleaned@meta.data <- seurat.obj.f_cleaned@meta.data %>%
  mutate(type = dplyr::case_when(
    class %in% c("PT", "DTL", "TAL", "DCT", "CNT", "PC", "ICA", "ICB", "PODO", "PEC", "URO") ~ "EPITHELIAL",
    class == "EC" ~ "ENDOTHELIAL",
    class %in% c("FIB", "MES", "VSMC") ~ "STROMAL",
    class == "IMMUNE" ~ "IMMUNE",
  ))

type_order <- c("EPITHELIAL", "ENDOTHELIAL", "STROMAL", "IMMUNE")
seurat.obj.f_cleaned@meta.data$type <- factor(seurat.obj.f_cleaned@meta.data$type, levels = type_order)

DimPlot(seurat.obj.f_cleaned, group.by = "type", label = TRUE, label.size = 4) +
  theme(plot.title = element_text(hjust = 0.5)) +
  ggtitle(paste0("Type Annotation for ", kidney1))
```
### Note: Absence of immune cells?

```{r}
type_order <- c("EPITHELIAL", "ENDOTHELIAL", "STROMAL", "IMMUNE")
seurat.obj.f_cleaned@meta.data$type <- factor(seurat.obj.f_cleaned@meta.data$type, levels = type_order)

markers.to.plot3 <- c("Ank3",         # PTS1, PTS2, PTS3, DTL, TAL1, TAL2, MD, DCT1, DCT2, CNT, PC, ICA, ICB, PODO, PEC, URO
                      "Flt1",         # EC
                      "Pdgfrb",       # FIB, MES, VSMC
                      "Ptprc"         # LYMPHO, MACRO
)

DotPlot(seurat.obj.f_cleaned,
features = markers.to.plot3,
group.by = "type",
dot.scale = 8,
dot.min = 0,
scale.max = 100,
scale.min = 0,
col.min = -2.5,
col.max = 2.5) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  coord_flip() +
  labs(x = "Gene Features", y = "Clusters") +  
  theme(plot.title = element_text(hjust = 0.5)) +
  ggtitle(paste0("Type Annotation for ", kidney1))
```

# Proportion Plots
```{r}
t1 <- table(seurat.obj.f_cleaned@meta.data$subclass)
#t1

prop.t1 <- prop.table(t1) 
#prop.t1

t2 <- as.data.frame(t1)
colnames(t2) <- c('Cell_type', 'Frequency')

# Original plot

f1a <- ggplot(t2, aes(fill=Cell_type, y=Frequency, x=1)) + 
  geom_bar(position="fill", stat = "identity", fun.y = "mean", colour="black") +
  theme_classic() +
  theme(axis.line = element_line(size = 1, colour = "black"),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        text = element_text(size=20),
        axis.line.x = element_blank()) +
  labs(fill = "Cell Type") +
  xlab(NULL) + ggtitle("Subclass")

t1 <- table(seurat.obj.f_cleaned@meta.data$class)
#t1

prop.t1 <- prop.table(t1) 
#prop.t1

t2 <- as.data.frame(t1)
colnames(t2) <- c('Cell_type', 'Frequency')

# Original plot

f1b <- ggplot(t2, aes(fill=Cell_type, y=Frequency, x=1)) + 
  geom_bar(position="fill", stat = "identity", fun.y = "mean", colour="black") +
  theme_classic() +
  theme(axis.line = element_line(size = 1, colour = "black"),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        text = element_text(size=20),
        axis.line.x = element_blank()) +
  labs(fill = "Cell Type") +
  xlab(NULL) + ggtitle("Class")

t1 <- table(seurat.obj.f_cleaned@meta.data$type)
#t1

prop.t1 <- prop.table(t1) 
#prop.t1

t2 <- as.data.frame(t1)
colnames(t2) <- c('Cell_type', 'Frequency')

# Original plot

f1c <- ggplot(t2, aes(fill = Cell_type, y = Frequency, x = 1)) + 
  geom_bar(position = "fill", stat = "identity", colour = "black") +
  theme_classic() +
  theme(
    axis.line = element_line(size = 1, colour = "black"),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    panel.border = element_blank(),
    panel.background = element_blank(),
    text = element_text(size = 20),
    axis.line.x = element_blank()
  ) +
  labs(fill = "Cell Type") +
  xlab(NULL) +
  ggtitle("Type")


f1a + f1b +f1c
```
### Note: How do I get this to look like the manual? Each bar seems to take up the same amount of space

# QC Reports: Filtering Progression Sankey Plot

```{r}
Start_name <- paste0(as.character(Start), " cells")
Filter_1_name <- paste0(as.character(Filter_1), " cells")
Filter_2_name <- paste0(as.character(Filter_2), " cells")
Filter_3_name <- paste0(as.character(Filter_3), " cells")


df_paths2 <- data.frame(
  group = c("Pass_All", "Filtered_at_3", "Filtered_at_2", "Filtered_at_1"),
  CellRanger = c(Start_name,     Start_name,       Start_name,        Start_name),
  nFeature = c(Filter_1_name,  Filter_1_name,    Filter_1_name,     paste0(as.character(Start-Filter_1), " cells removed")),
  doubletFinder = c(Filter_2_name,  Filter_2_name,   paste0(as.character(Filter_1-Filter_2), " cells removed"),    NA),
  ManualDoublets = c(Filter_3_name, paste0(as.character(Filter_2-Filter_3), " cells removed"),      NA,       NA),
  value = c(Start,        Filter_2-Filter_3,          Filter_1-Filter_2,           Start-Filter_1)
)


df_paths2b <- data.frame(
  group = c("Pass_All", "Filtered_at_3", "Filtered_at_2", "Filtered_at_1"),
  CellRanger = c(Start_name, Start_name, Start_name, Start_name),
  nFeature = c(Filter_1_name, Filter_1_name, Filter_1_name, paste0(as.character(Start - Filter_1), " cells removed")),
  doubletFinder = c(Filter_2_name, Filter_2_name, paste0(as.character(Filter_1 - Filter_2), " cells removed"), NA),
  ManualDoublets = c(Filter_3_name, paste0(as.character(Filter_2 - Filter_3), " cells removed"), NA, NA),
  Final = c(Filter_3_name , NA, NA, NA),
  value = c(Start, Filter_2 - Filter_3, Filter_1 - Filter_2, Start - Filter_1)
)

# Convert to long format suitable for ggsankey
df_sankey <- df_paths2b   %>%
  make_long(CellRanger, nFeature, doubletFinder, ManualDoublets, Final, value = "value")

# make removed nodes gray
orig_levels <- levels(factor(df_sankey$node))
orig_palette <- scales::hue_pal()(length(orig_levels))
names(orig_palette) <- orig_levels

# Override only "removed" nodes to gray
custom_palette <- orig_palette
custom_palette[grepl("removed", names(custom_palette))] <- "gray"

# change font size of removed nodes
df_sankey <- df_sankey %>%
  filter(!is.na(node))%>%
  mutate(is_removed = grepl("removed", node),
         label_size = ifelse(is_removed, 3, 4))

# specify order
df_sankey$node <- factor(df_sankey$node, levels = rev(unique(df_sankey$node)))
df_sankey$next_node <- factor(df_sankey$next_node, levels = rev(unique(df_sankey$node)))

# Do the same for next_node
df_sankey$next_node <- factor(df_sankey$next_node, levels = levels(df_sankey$node))


# Plot
ggplot(df_sankey, aes(x = x,
                      next_x = next_x,
                      node = node,
                      next_node = next_node,
                      value = value,
                      fill = factor(node))) +
  geom_sankey(flow.alpha = 0.7, node.color = "black") +
  geom_sankey_label(aes(label = node, size = label_size), color = "white", show.legend = FALSE) +
  scale_fill_manual(values = custom_palette) +   
  scale_color_identity() +
  scale_size_identity() +
  theme_classic() +
labs(
  title = paste0("Filtering Progression Sankey Plot for ", kidney1),
  x = ""
) +
  theme(
    # Remove Y-axis
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),
    axis.line.y = element_blank(),
    axis.title.y = element_blank(),
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),

    # Remove X-axis lines
    axis.line.x = element_blank(),
    axis.ticks.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),

    # Enlarge text
    axis.text.x = element_text(size = 14),
    axis.title.x = element_text(size = 16),
    plot.title = element_text(size = 18, face = "bold", hjust = 0.5),

    # Hide legend
    legend.position = "none"
  )
```

# Subclass cell types that were not annotated
```{r}
list1 <- unique(seurat.obj.f_cleaned$subclass)
  
list2 <- unique(markers_long$Subclass)

missing_subclass <- setdiff(list2, list1)

missing_subclass
```

### TALA, MD, PEC, URO, LYMPH, MES, VSMC, LYMPHO, MACRO were not annotated in this sample

### Note: Had trouble w/ saving Seurat object, will receive help in Jan 9th meeting