---
title: "qc_dataset1"
author: "jeff karnsomprot"
date: "2025-12-19"
output:
  html_document:
    toc: yes
    toc_depth: 4
    toc_float:
      collapsed: true
      number_sections: yes
    theme: journal
    df_print: paged
    code_folding: hide
    highlight: pygments
editor_options: 
  chunk_output_type: inline
---

# LOADING PACKAGES

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE}

if (!require("dplyr")) {install.packages("dplyr"); require("dplyr")}
if (!require("Seurat")) {install.packages("Seurat"); require("Seurat")}
if (!require("patchwork")) {install.packages("patchwork"); require("patchwork")}
if (!require("cowplot")) {install.packages("cowplot"); require("cowplot")}
if (!require("ggpubr")) {install.packages("ggpubr"); require("ggpubr")}
if (!require("plotly")) {install.packages("plotly"); require("plotly")}
if (!require("knitr")) {install.packages("knitr"); require("knitr")}
if (!require("htmlwidgets")) {install.packages("htmlwidgets"); require("htmlwidgets")}
if (!require("here")) {install.packages("here"); require("here")}
if (!require("EnhancedVolcano")) {BiocManager::install('EnhancedVolcano'); require("EnhancedVolcano")} # volcano plot
if (!require("tidyverse")) {install.packages("tidyverse"); require("tidyverse")} # for titying up data
if (!require("RColorBrewer")) {install.packages("RColorBrewer"); require("RColorBrewer")} # for color brewer
if (!require("sctransform")) {install.packages("sctransform"); require("sctransform")} # for data normalization
if (!require("glmGamPoi")) {BiocManager::install('glmGamPoi'); require("glmGamPoi")} # for data normalization, sctransform
if (!require("openxlsx")) {install.packages("openxlsx"); require("openxlsx")} # to save .xlsx files
if (!require("SoupX")) {install.packages("SoupX"); require("SoupX")}
if (!require("DoubletFinder")) {BiocManager::install('DoubletFinder'); require("DoubletFinder")}
if (!require("qs")) {install.packages("qs"); require("qs")}


set.seed((12345))
here()

```



# Removing Ambient RNA with SoupX

The current strategy that we use to remove doublets relies on [(Young and Behjati, 2020)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7763177/)

In order for this code to work you need both the *raw_feature_bc_matrix* in addition to the *filteredfiltered_feature_bc_matrix*. 

This pipeline was written by former visiting medical student [(Jeremiah Reyes)](https://twitter.com/imyourNephBro). 


```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center'}

#Loading Dataset
tod <- Read10X_h5(here("datasets", "1_raw_feature_bc_matrix.h5")) # change 
toc <- Read10X_h5(here("datasets", "1_filtered_feature_bc_matrix.h5")) # change 

sc = SoupChannel(tod,toc)

#Make the Seurat object from the filtered control data
SO <- Read10X_h5(here("datasets","1_filtered_feature_bc_matrix.h5")) #Change 
SO <- CreateSeuratObject(counts = SO, project = "control_ace2")  #Change

#Cluster the cells with Seurat
SO <- SCTransform(SO, verbose = F)
SO <- RunPCA(SO, verbose = F)
SO <- RunUMAP(SO, dims = 1:30, verbose = F)
SO <- FindNeighbors(SO, dims = 1:30, verbose = F)
SO <- FindClusters(SO, verbose = T)
 
meta <- SO@meta.data
umap <- SO@reductions$umap@cell.embeddings
clusters <- setNames(meta$seurat_clusters, rownames(meta))

#Sanity Check
length(clusters) #should be equal to nrow(sc$metaData)
nrow(sc$metaData)

sc <- setClusters(sc, clusters)
sc <- setDR(sc, umap)

#Estimate rho
sc = autoEstCont(sc)

#Clean the data
SO_out = adjustCounts(sc)
 
#Create a new Seurat Object out of the cleaned data
seurat.obj <- CreateSeuratObject(SO_out)
```

# Remove Doublets with the Doublet Detector Package

The output of the SoupX chunck feeds right into the doublet detector pipline which is based on [(McGinnis et al., 2019)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6853612/).  

```{r, echo=TRUE, error=FALSE, fig.align= 'center', message=FALSE, warning=FALSE}

VlnPlot(seurat.obj, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)

# Minimal QC and Filtering (low quality cells) to let doublet find doublets

seurat.obj.f <- subset(seurat.obj, nFeature_RNA > 500)

VlnPlot(seurat.obj.f, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)

seurat.obj.f

# Pre-process standard workflow
seurat.obj.f <- NormalizeData(object = seurat.obj.f)
seurat.obj.f <- FindVariableFeatures(object = seurat.obj.f)
seurat.obj.f <- ScaleData(object = seurat.obj.f)
seurat.obj.f <- RunPCA(object = seurat.obj.f)
ElbowPlot(seurat.obj.f, ndims = 40)

# PCs between 15-20
seurat.obj.f <- FindNeighbors(object = seurat.obj.f, dims = 1:30)
seurat.obj.f <- FindClusters(object = seurat.obj.f, resolution = 0.03)
seurat.obj.f <- RunUMAP(object = seurat.obj.f, dims = 1:30)
DimPlot(seurat.obj.f, reduction = "umap")

```

## paramSweep

A very time-consuming step.

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center', results=F}

# Calculate each combination of pN and pK
sweep.res.list_seurat.obj.f <- paramSweep(seurat.obj.f, PCs = 1:20, sct = FALSE) 
```

## summarizeSweep

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center'}

#Summarize each combination of pN and pK
sweep.stats_seurat.obj.f <- summarizeSweep(sweep.res.list_seurat.obj.f, GT = FALSE) 

#Select the pK that corresponds to max bcmvn to optimize doublet detection
bcmvn_seurat.obj.f <- find.pK(sweep.stats_seurat.obj.f)
pK <- bcmvn_seurat.obj.f %>% 
  filter(BCmetric == max(BCmetric)) %>%
  select(pK) 

#See pK in the Values Environment
pK <- as.numeric(as.character(pK[[1]]))

```

## summarizeSweep

```{r , echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center'}
 
# Homotypic Doublet Proportion Estimate -------------------------------------------------------------------------------------
annotations <- seurat.obj.f@meta.data$seurat_clusters  
 
homotypic.prop <- modelHomotypic(annotations)           
homotypic.prop
 
# 10X Multiplet Rate Table (the doublet ratio is # of cells recovered divided by 125000) https://kb.10xgenomics.com/hc/en-us/articles/360001378811-What-is-the-maximum-number-of-cells-that-can-be-profiled-
 
nrow(seurat.obj.f@meta.data)

nExp_poi <- round(nrow(seurat.obj.f@meta.data) # To calculate cell number
                  /125000              # To calculate the doublet ratio
                  *nrow(seurat.obj.f@meta.data))
nExp_poi

nExp_poi_adj <- round(nExp_poi*(1-homotypic.prop))
 
```

## doubletFinder

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center'}

seurat.obj.f_doublets <- doubletFinder(seurat.obj.f,
                        PCs = 1:20,
                        pN = 0.25,
                        pK = pK,
                        nExp = nExp_poi_adj,
                        reuse.pANN = FALSE, sct = FALSE)
colnames(seurat.obj.f_doublets@meta.data)[6] <- "pANN"
colnames(seurat.obj.f_doublets@meta.data)[7] <- "DF.class"
head(seurat.obj.f_doublets@meta.data)
table(seurat.obj.f_doublets@meta.data$DF.class)

DimPlot(seurat.obj.f_doublets, group.by = "DF.class")

VlnPlot(seurat.obj.f_doublets, "nFeature_RNA", group.by = "DF.class")

VlnPlot(seurat.obj.f_doublets, "nCount_RNA", group.by = "DF.class")

```

## Subset Singlets

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center'}

seurat.obj.f_singlets <- subset(seurat.obj.f_doublets, DF.class == "Singlet")
seurat.obj.f_singlets
DimPlot(seurat.obj.f_singlets, reduction = "umap")



```

### Remove Mitochondrial Genes

Because it isn't clear what the percentage of mitochondrial genes means in a single-*nucleus* RNAseq dataset, we will take out all the mitochondrial genes to remove thier effect on clustering. However, before we remove them from the dataset, we will calculate their values and *stash* them in the metadata. 

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center'}

seurat.obj.f_singlets <- seurat.obj.f_singlets[!grepl("^mt-", rownames(seurat.obj.f_singlets)), ]

#Mito Sanity Check
counts <- GetAssayData(seurat.obj.f_singlets, assay = "RNA")
mito.genes <- grep(pattern = "^mt-", x = rownames(x = counts), value = TRUE) 
mito.genes #should be zero

DimPlot(seurat.obj.f_singlets, reduction = "umap", label = T)

```

# Initial Cluster Identification

```{r}

ElbowPlot(seurat.obj.f_singlets)
seurat.obj.f_singlets <- SCTransform(seurat.obj.f_singlets) %>%
    RunPCA() %>%
    FindNeighbors(dims = 1:50) %>%
    FindClusters(resolution = 2) %>%
    RunUMAP(dims = 1:50)


DimPlot(seurat.obj.f_singlets, reduction = "umap", label = TRUE ) + 
  theme(plot.title = element_text(hjust = 0.5)) +
  ggtitle(paste0("Non-annotated Clusters for dataset1"))

```

Generic Pipeline for cluster identification

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center', fig.width = 7, fig.height= 8}

seurat.obj.f_singlets.markers <- FindAllMarkers(seurat.obj.f_singlets, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)

seurat.obj.f_singlets.markers %>%
    group_by(cluster) %>%
    top_n(n = 5, wt = avg_log2FC) -> top5

DoHeatmap(seurat.obj.f_singlets, features = top5$gene) + NoLegend()

seurat.obj.f_singlets.markers %>%
    group_by(cluster) %>%
    top_n(n = 2, wt = avg_log2FC) -> top2

VlnPlot(seurat.obj.f_singlets, 
        features = unique(top2$gene),
        stack = TRUE, 
        flip = TRUE,
        pt.size = 0)+
        NoLegend()

```


## Multidimensional Dotplot

```{r}

markers.to.plot1 <- c("Lrp2",         # PT

                      "Slc5a12",      # PT-S1

                      "Slc13a3",      # PT-S2

                      "Slc16a9",      # PT-S3

                      "Epha7",        # dTL
                      
                      "Cryab",         # dTL

                      "Slc12a1",      # TAL

                      "Cldn10",       # TAL

                      "Cldn16",       # TAL

                      "Nos1",         # MD

                      "Slc12a3",      # DCT

                      "Pvalb",        # DCT1

                      "Slc8a1",       # DCT2, CNT

                      "Aqp2",         # PC

                      "Slc4a1",       # IC-A

                      "Slc26a4",      # IC-B

                      "Upk1b",        # Uro

                      "Ncam1",        # PEC

                      "Pdgfrb",       # Perivascular

                      "Piezo2",       # Mesangial

                      "Pdgfra",       # Fib

                      "Acta2",        # Mural

                      "Nphs1",        # Podo

                      "Ptprc",        # Immune

                      "Cd74",         # Macrophage
                      
                      "Kdr",          # Capillary Endo
                      
                      "Flt1",
                      
                      "Emcn"

                      )


 

DotPlot(seurat.obj.f_singlets,

features = markers.to.plot1,group.by = "seurat_clusters",

dot.scale = 8,

dot.min = 0,

scale.max = 100,

scale.min = 0,

col.min = -2.5,

col.max = 2.5) +

  coord_flip() +

  theme_classic() +

  theme(axis.line = element_line(size = 1, colour = "black"),

        axis.ticks.x = element_blank(),

        panel.grid.major.x = element_blank(),

        panel.grid.minor.x = element_blank(),

        panel.border = element_blank(),

        panel.background = element_blank(),

        text = element_text(size=20),

        axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(x = "Gene Features", y = "Clusters")
  

```

```{r}

DimPlot(seurat.obj.f_singlets)
DefaultAssay(seurat.obj.f_singlets) <- "RNA"

```

## Assign names to Clusters

```{r}

# Apply names to the clusters by subclass
seurat.obj.f_singlets@meta.data <- seurat.obj.f_singlets@meta.data %>%
  mutate(subclass = dplyr::case_when(

    
    seurat_clusters == 0  ~ "PTS2",
    seurat_clusters == 1  ~ "EC",
    seurat_clusters == 2  ~ "PTS1",
    seurat_clusters == 3  ~ "PTS2",
    seurat_clusters == 4  ~ "PTS2",
    seurat_clusters == 5  ~ "TALB",
    seurat_clusters == 6  ~ "PTS1",
    seurat_clusters == 7  ~ "PTS3",
    seurat_clusters == 8  ~ "DCT1",     
    seurat_clusters == 9  ~ "TALA",
    seurat_clusters == 10 ~ "EC",
    seurat_clusters == 11 ~ "CNT",
    seurat_clusters == 12 ~ "PTS2",
    seurat_clusters == 13 ~ "FIB",
    seurat_clusters == 14 ~ "PC",
    seurat_clusters == 15 ~ "PTS1",
    seurat_clusters == 16 ~ "DTL",
    seurat_clusters == 17 ~ "PTS3",
    seurat_clusters == 18 ~ "DTL",
    seurat_clusters == 19 ~ "PC",
    seurat_clusters == 20 ~ "DCT2",
    seurat_clusters == 21 ~ "ICA",
    seurat_clusters == 22 ~ "DTL",
    seurat_clusters == 23 ~ "PTS1",
    seurat_clusters == 24 ~ "ICB",
    seurat_clusters == 25 ~ "DOUBLET",
    seurat_clusters == 26 ~ "MES",
    seurat_clusters == 27 ~ "URO",
    seurat_clusters == 28 ~ "CNT",
    seurat_clusters == 29 ~ "MD",
    seurat_clusters == 30 ~ "PODO",
    seurat_clusters == 31 ~ "MACRO",
    
    TRUE ~ NA_character_
))


subclass_order <- c(
  "PTS1", "PTS2", "PTS3", "DTL", "TALA", "TALB",
  "MD", "DCT1", "DCT2", "DCT3", "CNT", "PC",
  "ICA", "ICB", "PEC", "URO",
  "FIB", "MES","PODO", "MACRO", "LYMPH","EC","DOUBLET"
)


seurat.obj.f_singlets@meta.data$subclass <- factor(
  seurat.obj.f_singlets@meta.data$subclass,
  levels = subclass_order
)

Idents(seurat.obj.f_singlets) <- seurat.obj.f_singlets@meta.data$subclass

DimPlot(seurat.obj.f_singlets, reduction = "umap", label = TRUE, label.size = 4) +
  theme(plot.title = element_text(hjust = 0.5)) +
  ggtitle(paste0("Subclass Annotation for dataset1"))

```


# Subset Doublets

```{r}

seurat.obj.f_cleaned <- subset(seurat.obj.f_singlets, subset = subclass == "DOUBLET", invert = TRUE)

# Change the order of the identities on the x-axis to create a diagonal on the dotplot, going from top to bottom of the "naming kidney cells" table

seurat.obj.f_cleaned <- SCTransform(seurat.obj.f_cleaned) %>%
    RunPCA() %>%
    FindNeighbors(dims = 1:50) %>%
    FindClusters(resolution = 2) %>%
    RunUMAP(dims = 1:50)


```

# Assign Names to New Clean Dataset

```{r}

seurat.obj.f_cleaned@meta.data <- seurat.obj.f_cleaned@meta.data %>%
  mutate(subclass = dplyr::case_when(

    seurat_clusters == 0  ~ "PTS2",
    seurat_clusters == 1  ~ "PTS1",
    seurat_clusters == 2  ~ "EC",
    seurat_clusters == 3  ~ "PTS2",
    seurat_clusters == 4  ~ "TALB",
    seurat_clusters == 5  ~ "PTS3",
    seurat_clusters == 6  ~ "PTS1",
    seurat_clusters == 7  ~ "DCT2",
    seurat_clusters == 8  ~ "PTS2",
    seurat_clusters == 9  ~ "TALA",
    seurat_clusters == 10 ~ "EC",
    seurat_clusters == 11 ~ "CNT",
    seurat_clusters == 12 ~ "FIB",
    seurat_clusters == 13 ~ "PC",
    seurat_clusters == 14 ~ "PTS1",
    seurat_clusters == 15 ~ "PC",
    seurat_clusters == 16 ~ "DTL",
    seurat_clusters == 17 ~ "DTL",
    seurat_clusters == 18 ~ "PTS3",
    seurat_clusters == 19 ~ "DCT2",
    seurat_clusters == 20 ~ "DTL",
    seurat_clusters == 21 ~ "ICA",
    seurat_clusters == 22 ~ "PTS1",
    seurat_clusters == 23 ~ "ICB",
    seurat_clusters == 24 ~ "VSMC",
    seurat_clusters == 25 ~ "URO",
    seurat_clusters == 26 ~ "CNT",
    seurat_clusters == 27 ~ "MD",
    seurat_clusters == 28 ~ "DTL",
    seurat_clusters == 29 ~ "PODO",
    seurat_clusters == 30 ~ "MACRO",
    
    TRUE ~ NA_character_
))

# Specify the order you want for the factor levels
subclass_order <- c(
  "PTS1", "PTS2", "PTS3", "DTL", "TALA", "TALB",
  "MD", "DCT1", "DCT2", "CNT", "PC",
  "ICA", "ICB", "URO", "VSMC",
  "FIB", "PODO", "MACRO", "EC"
)

# Apply factor levels
seurat.obj.f_cleaned@meta.data$subclass <- factor(
  seurat.obj.f_cleaned@meta.data$subclass,
  levels = subclass_order
)

# Set Idents
Idents(seurat.obj.f_cleaned) <- seurat.obj.f_cleaned@meta.data$subclass

# UMAP Plot
DimPlot(seurat.obj.f_cleaned, reduction = "umap", label = TRUE, label.size = 4) +
  theme(plot.title = element_text(hjust = 0.5)) +
  ggtitle("Subclass Annotation for dataset1 (cleaned)")


DimPlot(seurat.obj.f_cleaned, reduction = "umap", label = TRUE ) + 
  theme(plot.title = element_text(hjust = 0.5)) +
  ggtitle(paste0("Non-annotated Clusters for dataset1"))

DotPlot(seurat.obj.f_cleaned,

features = markers.to.plot1,

dot.scale = 8,

dot.min = 0,

scale.max = 100,

scale.min = 0,

col.min = -2.5,

col.max = 2.5) +

  coord_flip() +

  theme_classic() +

  theme(axis.line = element_line(size = 1, colour = "black"),

        axis.ticks.x = element_blank(),

        panel.grid.major.x = element_blank(),

        panel.grid.minor.x = element_blank(),

        panel.border = element_blank(),

        panel.background = element_blank(),

        text = element_text(size=20),

        axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(x = "Gene Features", y = "Clusters")


```

## By class

```{r}

# Change the specificty of the cells to a lower level and make another UMAP
seurat.obj.f_cleaned@meta.data <- seurat.obj.f_cleaned@meta.data %>%
  mutate(class = dplyr::case_when(
    subclass %in% c("PTS1", "PTS2", "PTS3") ~ "PT",
    subclass == "DTL" ~ "DTL",
    subclass %in% c("TALA", "TALB", "MD") ~ "TAL",
    subclass %in% c("DCT1", "DCT2") ~ "DCT",
    subclass == "CNT" ~ "CNT",
    subclass == "PC" ~ "PC",
    subclass == "ICA" ~ "ICA",
    subclass == "ICB" ~ "ICB",
    subclass == "PODO" ~ "PODO",
    subclass == "PEC" ~ "PEC",
    subclass == "URO" ~ "URO",
    subclass %in% c("EC", "ECG", "ECP", "LYMPH") ~ "EC",
    subclass == "FIB" ~ "FIB",
    subclass == "MES" ~ "MES",
    subclass == "VSMC" ~ "VSMC",
    subclass %in% c("LYMPHO", "MACRO") ~ "IMMUNE",
  ))

DimPlot(seurat.obj.f_cleaned, group.by = "class", label = TRUE, label.size = 4) +
   theme(plot.title = element_text(hjust = 0.5)) +
   ggtitle(paste0("Class Annotation for dataset 1"))
   
```

```{r}

class_order <- c("PT", "DTL", "TAL", "DCT", "CNT", "PC", "ICA", "ICB", "PODO", "PEC", "URO", "EC", "FIB", "MES", "IMMUNE")
seurat.obj.f_cleaned@meta.data$class <- factor(seurat.obj.f_cleaned@meta.data$class, levels = class_order)

markers.to.plot2 <- c("Lrp2",         # PTS1, PTS2, PTS3
                      "Cryab",        # DTL
                      "Slc12a1",      # TAL1, TAL2, MD
                      "Slc12a3",      # DCT1, DCT2
                      "Slc8a1",       # CNT
                      "Aqp2",         # PC
                      "Kit",          # ICA
                      "Slc26a4",      # ICB
                      "Upk1b",        # URO
                      "Piezo2",       # MES
                      "Pdgfra",       # FIB
                      "Pdgfrb",       # FIB, MES, VSMC
                      "Nphs1",        # PODO
                      "Ncam1",        # PEC
                      "Ptprc",        # LYMPHO, MACRO
                      "Flt1"       # EC
)

DotPlot(seurat.obj.f_cleaned,

features = markers.to.plot2,

dot.scale = 8,

dot.min = 0,

scale.max = 100,

scale.min = 0,

col.min = -2.5,

col.max = 2.5) +

  coord_flip() +

  theme_classic() +

  theme(axis.line = element_line(size = 1, colour = "black"),

        axis.ticks.x = element_blank(),

        panel.grid.major.x = element_blank(),

        panel.grid.minor.x = element_blank(),

        panel.border = element_blank(),

        panel.background = element_blank(),

        text = element_text(size=20),

        axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(x = "Gene Features", y = "Clusters")



```

## By type

```{r}


# Change the specificty of the cells to a further lower level and make another UMAP
seurat.obj.f_cleaned@meta.data <- seurat.obj.f_cleaned@meta.data %>%
  mutate(type = dplyr::case_when(
    class %in% c("PT", "DTL", "TAL", "DCT", "CNT", "PC", "ICA", "ICB", "PODO", "PEC", "URO") ~ "EPITHELIAL",
    class == "EC" ~ "ENDOTHELIAL",
    class %in% c("FIB", "MES", "VSMC") ~ "STROMAL",
    class == "IMMUNE" ~ "IMMUNE",
  ))

type_order <- c("EPITHELIAL", "ENDOTHELIAL", "STROMAL", "IMMUNE")
seurat.obj.f_cleaned@meta.data$type <- factor(seurat.obj.f_cleaned@meta.data$type, levels = type_order)

DimPlot(seurat.obj.f_cleaned, group.by = "type", label = TRUE, label.size = 4) +
  theme(plot.title = element_text(hjust = 0.5)) +
  ggtitle(paste0("Type Annotation for dataset 1"))

```




type_order <- c("EPITHELIAL", "ENDOTHELIAL", "STROMAL", "IMMUNE")
seurat.obj.f_cleaned@meta.data$type <- factor(seurat.obj.f_cleaned@meta.data$type, levels = type_order)

markers.to.plot3 <- c("Ank3",         # PTS1, PTS2, PTS3, DTL, TAL1, TAL2, MD, DCT1, DCT2, CNT, PC, ICA, ICB, PODO, PEC, URO
                      "Flt1",         # EC
                      "Pdgfrb",       # FIB, MES, VSMC
                      "Ptprc"         # LYMPHO, MACRO
)


DotPlot(seurat.obj.f_cleaned,

features = markers.to.plot3,group.by = "type",

dot.scale = 8,

dot.min = 0,

scale.max = 100,

scale.min = 0,

col.min = -2.5,

col.max = 2.5) +

  coord_flip() +

  theme_classic() +

  theme(axis.line = element_line(size = 1, colour = "black"),

        axis.ticks.x = element_blank(),

        panel.grid.major.x = element_blank(),

        panel.grid.minor.x = element_blank(),

        panel.border = element_blank(),

        panel.background = element_blank(),

        text = element_text(size=20),

        axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(x = "Gene Features", y = "Clusters")





# Proportion plot



t1 <- table(seurat.obj.f_cleaned@meta.data$subclass)

prop.t1 <- prop.table(t1) 

t2 <- as.data.frame(t1)
colnames(t2) <- c('Cell_type', 'Frequency')

f1a <- ggplot(t2, aes(fill=Cell_type, y=Frequency, x=1)) + 
  geom_bar(position="fill", stat = "identity", fun.y = "mean", colour="black") +
  theme_classic() +
  theme(axis.line = element_line(size = 1, colour = "black"),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        text = element_text(size=20),
        axis.line.x = element_blank()) +
  labs(fill = "Cell Type") +
  xlab(NULL) + ggtitle("Subclass")

t1 <- table(seurat.obj.f_cleaned@meta.data$class)

prop.t1 <- prop.table(t1) 

t2 <- as.data.frame(t1)
colnames(t2) <- c('Cell_type', 'Frequency')


f1b <- ggplot(t2, aes(fill=Cell_type, y=Frequency, x=1)) + 
  geom_bar(position="fill", stat = "identity", fun.y = "mean", colour="black") +
  theme_classic() +
  theme(axis.line = element_line(size = 1, colour = "black"),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        text = element_text(size=20),
        axis.line.x = element_blank()) +
  labs(fill = "Cell Type") +
  xlab(NULL) + ggtitle("Class")

t1 <- table(seurat.obj.f_cleaned@meta.data$type)

prop.t1 <- prop.table(t1) 

t2 <- as.data.frame(t1)
colnames(t2) <- c('Cell_type', 'Frequency')


f1c <- ggplot(t2, aes(fill = Cell_type, y = Frequency, x = 1)) + 
  geom_bar(position = "fill", stat = "identity", colour = "black") +
  theme_classic() +
  theme(
    axis.line = element_line(size = 1, colour = "black"),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    panel.border = element_blank(),
    panel.background = element_blank(),
    text = element_text(size = 20),
    axis.line.x = element_blank()
  ) +
  labs(fill = "Cell Type") +
  xlab(NULL) +
  ggtitle("Type")


f1a + f1b +f1c


# Saving Qq File

```{r}

qsave(seurat.obj.f_cleaned, file = here("jk_code", "CleanDS1.rds"))

```